reference-delegate-explainer.md

```diff
@@ -0,0 +1,534 @@
+                                            aria-activedescendant: option-2">
+    <div id="real-listbox" role="listbox">
+      <div id="option-1" role="option">Option 1</div>
+      <div id="option-2" role="option">Option 2</div>
+    </div>
+  </template>
+</fancy-listbox>
+```
+
+Declaring the mappings using the JavaScript API would look like:
+```js
+this.shadowRoot_.referenceDelegateMap['aria-controls'] = 'real-listbox';
+this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = 'option-2';
+```
+
+> Note: the syntax of `shadowrootreferencedelegatemap` is based on the [`exportparts`](https://drafts.csswg.org/css-shadow-parts/#exportparts-attr) attribute that contains a comma-separated map of part names.
+
+#### Combining `referenceDelegate` and `referenceDelegateMap`
+
+In the case where both attributes are specified, `referenceDelegateMap` takes priority for individual attributes, and `referenceDelegate` acts as the fallback for attributes that are not specified.
+
+In the example below, `"real-listbox"` is the delegate for all attributes _except_ `aria-activedescendant`, which is delegated to `"option-2"`.
+
+```html
+<input role="combobox"
+       aria-controls="fancy-listbox"
+       aria-activedescendant="fancy-listbox" />
+<fancy-listbox id="fancy-listbox">
+  <template shadowrootmode="open"
+            shadowrootreferencedelegate="real-listbox"
+            shadowrootreferencedelegatemap="aria-activedescendant: option-2">
+    <div id="real-listbox" role="listbox">
+      <div id="option-1" role="option">Option 1</div>
+      <div id="option-2" role="option">Option 2</div>
+    </div>
+  </template>
+</fancy-listbox>
+```
+
+#### Delegating to multiple elements
+
+Some attributes such as `aria-labelledby`, `aria-describedby`, etc. support multiple targets. Using `referenceDelegateMap` with those attributes support a space-separated list of IDs.
+
+This example shows a `<description-with-tooltip>` component that contains a "More Info" button to show the tooltip but is not intended to be included in the description text. It delegates `aria-describedby: message tooltip` to forward to only the content that should be included in the description text.
+
+```html
+<input aria-describedby="description-with-tooltip" />
+<!--
+  The resulting description text is: 
+  "Inline description text. Tooltip with more information."
+-->
+<description-with-tooltip id="description-with-tooltip">
+  <template shadowrootmode="closed"
+            shadowrootreferencedelegatemap="aria-describedby: message tooltip">
+    <div>
+      <span id="message">
+        Inline description text.
+      </span>
+      <button onmouseover="showTooltip()" onmouseout="hideTooltip()">
+        More Info
+      </button>
+      <div id="tooltip" role="tooltip" style="display: none">
+        Tooltip with more information.
+      </div>
+    </div>
+  </template>
+</description-with-tooltip>
+```
```

Comment:
If I understand this proposal correctly, it does still suffer from the bottleneck effect, right? For example, in this case, the consumer can't decide whether to refer to the `message` _or_ the `tooltip` – it can only refer to both at the same time.

Maybe this is an acceptable tradeoff (and as you say elsewhere, similar proposals have been delayed due to scope creep), but maybe it's worth calling out as an explicit non-goal.

Reply:
Yes, it does suffer from the [bottleneck effect](https://alice.pages.igalia.com/blog/how-shadow-dom-and-accessibility-are-in-conflict/#limitations-of-these-apis). However, I haven't been able to come up with a compelling example where the bottleneck effect is a limitation in real-world use.

The example in your comment isn't actually the bottleneck effect: the `description-with-tooltip` component _could_ provide some API to let the consumer pick whether to include the `message` or `tooltip` or both. The component would update its `referenceDelegateMap` based on that API.

The bottleneck effect would be if there were two separate elements outside of the `description-with-tooltip`, and one wanted its `aria-describedby` to be the `message`, and the other wanted its `aria-describedby` to be the `tooltip`. I can't think of a good reason why you'd do that...